File: lab10/src/by/bntu/fitr/minigame/domain/east/Dwarf.java
package by.bntu.fitr.minigame.domain.east;

import by.bntu.fitr.minigame.domain.Creature;
import by.bntu.fitr.minigame.domain.Role;
import by.bntu.fitr.minigame.domain.Weapon;

public class Dwarf extends Creature {
    private Weapon weapon = Weapon.AXE;
    private Role role = Role.TANK;

    public Dwarf(int health, int strength, int agility, int intelligence) {
        super(health, strength, agility, intelligence);
    }

    @Override
    protected int getWeaponDamage() {
        return weapon.getDamage();
    }

    @Override
    public int buff() {
        return role.getPercent() / 100 * getHealth();
    }

    @Override
    public String toString() {
        return "Dwarf{" + super.toString() +
                "weapon=" + weapon +
                ", role=" + role +
                '}';
    }
}

File: lab10kt/src/main/kotlin/by/bntu/fitr/povt/korotkevich/model/west/Dwarf.kt
package by.bntu.fitr.povt.korotkevich.model.west

import by.bntu.fitr.povt.korotkevich.model.Creature
import by.bntu.fitr.povt.korotkevich.model.Role
import by.bntu.fitr.povt.korotkevich.model.Weapon
import by.bntu.fitr.povt.korotkevich.model.buff

class Dwarf(override val health: Int = 200, override val strength: Int = 50,
            override val agility: Int = 10, override val intelligence: Int = 10,
            val role: Role = Role.TANK, val weapon: Weapon = Weapon.HAMMER) : Creature() {

    override val id: String = "Dwarf"

    override fun getSelfDamage() = weapon.damage + buff(role)

    override fun toString() = StringBuilder("Dwarf(").apply {
        append("${super.toString()}, role: $role, weapon: $weapon)")
    }.toString()
}